### PR TITLE
Add support for message attachments

### PR DESCRIPTION
**Pull request by jerith**
This pull request addresses #11.

I haven't given much thought at all to API design in the initial implementation here, I just hacked in some types and functions to get something more or less working.

### CODE REVIEW

**Comment by Leonidas-from-XIV** [comment]
Looking forward to reviewing it! Can you squash these two commits?

File: src/slack_notify.ml [hunk 1]
```diff
@@ -42,11 +42,15 @@ let message =
...
+let attachment =
+  let doc = "Attachment text" in
+  Cmdliner.Arg.(value & opt (some string) None & info ["attachment"] ~docv:"MSG" ~doc)
+
 let info =
   let doc = "Posts messages to Slack" in
   Cmdliner.Term.info "slack-notify" ~doc
 
-let execute token username channel icon_url icon_emoji msg =
+let execute token username channel icon_url icon_emoji atxt msg =
```
Comment (Leonidas-from-XIV) [inline src/slack_notify.ml]: `attachment_text` would be better or something, since `atxt` is a rather unexpressive name.

File: src/slacko.ml [hunk 2]
```diff
@@ -335,6 +335,64 @@ type file_obj = {
...
 
+type field_obj = {
+  title: string option [@default None];
+  value: string option [@default None];
+  short: bool option [@default None];
+} [@@deriving yojson { strict = false }]
+
+let field ?title ?value ?short () = {
+  title;
+  value;
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: Does it make sense to construct a `field_obj` without a value? Making the `value` mandatory would allow to avoid the `()` argument at maybe no adverse cost to the utility of the API.

File: src/slacko.ml [hunk 3]
```diff
@@ -335,6 +335,64 @@ type file_obj = {
...
+  short;
+}
+
+type attachment_obj = {
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: Can these really be all `None`?

**Comment by jerith** [comment]
Done.

File: src/slack_notify.ml [hunk 4]
```diff
@@ -66,16 +70,22 @@ let execute token username channel icon_url icon_emoji msg =
   let channel = Slacko.channel_of_string channel in
   let chat = Slacko.Channel channel in
   let msg = Slacko.message_of_string msg in
+  let attachments =
```
Comment (Leonidas-from-XIV) [inline src/slack_notify.ml]: Maybe it would make sense to implement our own `Option.map` to avoid this boilerplat-y match.

File: src/slack_notify.ml [hunk 5]
```diff
@@ -66,16 +70,22 @@ let execute token username channel icon_url icon_emoji msg =
...
     ?icon_emoji:(icon_emoji)
     ?icon_url:(icon_url)
+    ?attachments:(attachments)
     msg
   >|= (fun c ->
     print_endline @@ string_or_bust c)
   |> Lwt_main.run
 
-let execute_t = Cmdliner.Term.(pure execute $ token $ username $ channel $ icon_url $ icon_emoji $ message)
+let execute_t = Cmdliner.Term.(pure execute $ token $ username $ channel $ icon_url $ icon_emoji $ attachment $ message)
```
Comment (Leonidas-from-XIV) [inline src/slack_notify.ml]: At this point it would make sense to wrap this line (actually this should've happened before but as we're changing the line anyway).

File: src/slacko.ml [hunk 6]
```diff
@@ -335,6 +335,64 @@ type file_obj = {
...
+let attachment
+    ?fallback ?color ?pretext
+    ?author_name ?author_link ?author_icon
+    ?title ?title_link
+    ?text ?fields
+    ?image_url ?thumb_url
+    ?footer ?footer_icon
+    ?ts ?mrkdwn_in
+    () = {
+  fallback = (match fallback with Some _ -> fallback | None -> text);
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: The parens are not really necessary.

File: src/slacko.ml [hunk 7]
```diff
@@ -1075,8 +1133,14 @@ let chat_delete token ts chat =
     | #message_error as res -> res
     | _ -> `Unknown_error
 
+let jsonify_attachments = function
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: It would be potentially nicer to just have this operate on attachments directly and use an `Option.map` like function to conditionally dispatch a call to this function.

File: src/slacko.ml [hunk 8]
```diff
@@ -1086,6 +1150,7 @@ let chat_post_message token chat
     |> optionally_add "parse" parse
     |> optionally_add "icon_url" icon_url
     |> optionally_add "icon_emoji" icon_emoji
+    |> optionally_add "attachments" (jsonify_attachments attachments)
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: Use `@@` instead of the parens, to be consistent with the remaining code.

**Comment by Leonidas-from-XIV** [comment]
(Sorry for the noise, I rarely review on GitHub, therefore I am not super familiar with the process)

File: src/slacko.ml (hunk 2 above)
Comment (jerith) [inline src/slacko.ml]: Good point, I'll change that.

File: src/slacko.ml (hunk 3 above)
Comment (jerith) [inline src/slacko.ml]: I've successfully posted attachments containing only `fields` and only `text`, so there aren't any specific fields that are mandatory.

File: src/slack_notify.ml (hunk 4 above)
Comment (jerith) [inline src/slack_notify.ml]: It would. My focus was on getting the functionality working so I could get some feedback on the general approach, so there's probably a bunch of stuff like this that I can do fairly easily. :-)

File: src/slacko.ml (hunk 6 above)
Comment (jerith) [inline src/slacko.ml]: I put them in because the compiler thought the next field was part of the `None -> text` block, but that may have been a slightly different version of the expression. I'll check again and remove them if I can.

File: src/slacko.ml (hunk 7 above)
Comment (jerith) [inline src/slacko.ml]: Yes. This is one of the places where I didn't put much thought into the code beyond "does this work at all", so it could definitely use some more attention in general.

File: src/slacko.ml (hunk 8 above)
Comment (jerith) [inline src/slacko.ml]: Will do.

**Comment by jerith** [comment]
Thanks for the review, I'll push some updates a little later today.

Are you happy with the new types and functions for working with them? They're not particularly consistent with the rest of the Slacko API, but I don't really have any good ideas for cleaning that up.

File: src/slacko.ml (hunk 6 above)
Comment (jerith) [inline src/slacko.ml]: Yes, the parens are necessary for the pattern match. I can remove them by factoring that out into a function, though.

File: src/slack_notify.ml (hunk 4 above)
Comment (jerith) [inline src/slack_notify.ml]: Pulling in `maybe` from `slacko.ml` and writing a function to do the `[Slacko.attachment ~text:txt ()]` costs about as much space as it saves by not having the pattern match, and the end result is harder to read at a glance. I'll remove the unnecessary parens, though.

**Comment by jerith** [comment]
I think I've addressed all the comments in this round of review.

File: src/slacko.ml [hunk 9]
```diff
@@ -366,6 +366,8 @@ type attachment_obj = {
   mrkdwn_in: string list option [@default None];
 } [@@deriving yojson { strict = false }]
 
+let ifnone a b = match a with Some v -> Some v | None -> b
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: I'd name this `if_none` to match OCaml naming scheme. And maybe make this multi-line, since there is not much use in saving lines here.

File: src/slacko.ml (hunk 3 above)
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: If `short` is not specified, maybe we can make it default to `false`? Unfortunately the Slack API docs are unclear on what it defaults to or whether it is a tri-state boolean:

> An optional flag indicating whether the value is short enough to be displayed side-by-side with other values.

File: src/slacko.ml [hunk 10]
```diff
@@ -335,6 +335,66 @@ type file_obj = {
...
+  title;
+  title_link;
+  text;
+  fields;
+  image_url;
+  thumb_url;
+  footer;
+  footer_icon;
+  ts;
+  mrkdwn_in;
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: Where did you find `mrkdwn_in`? I couldn't locate it in the docs.

File: src/slacko.mli [hunk 11]
```diff
@@ -697,7 +737,7 @@ val channels_unarchive: token -> channel -> [ `Success | parsed_auth_error | cha
 val chat_delete: token -> timestamp -> chat -> [ `Success of chat_obj | parsed_auth_error | channel_error | message_error ] Lwt.t
 
 (** Sends a message to a channel. *)
-val chat_post_message: token -> chat -> ?username:string -> ?parse:string -> ?icon_url:string -> ?icon_emoji:string -> message -> [ `Success of chat_obj | parsed_auth_error | channel_error | archive_error | message_length_error | rate_error | bot_error ] Lwt.t
+val chat_post_message: token -> chat -> ?username:string -> ?parse:string -> ?icon_url:string -> ?icon_emoji:string -> ?attachments:attachment_obj list -> message -> [ `Success of chat_obj | parsed_auth_error | channel_error | archive_error | message_length_error | rate_error | bot_error ] Lwt.t
```
Comment (Leonidas-from-XIV) [inline src/slacko.mli]: This will also potentially emit a new error, `too_many_attachments`, so that should be added somewhere.

File: src/slacko.ml (hunk 10 above)
Comment (jerith) [inline src/slacko.ml]: It's buried in https://api.slack.com/docs/message-formatting#formatting_and_attachments for some reason. I only discovered it after spending way too much time trying to figure out why my attachments weren't being formatted.

Comment (Leonidas-from-XIV) [inline src/slacko.ml]: Oh, looks `mrkdwn` needs to be added somewhere as well. But not necessarily in this PR. Thanks for pointing it out!

File: src/slacko.ml [hunk 12]
```diff
@@ -1075,8 +1135,12 @@ let chat_delete token ts chat =
     | #message_error as res -> res
     | _ -> `Unknown_error
 
+let jsonify_attachments attachments =
+  `List (List.map (fun a -> attachment_obj_to_yojson a) attachments)
+  |> Yojson.Safe.to_string
+
 let chat_post_message token chat
-  ?username ?parse ?icon_url ?icon_emoji text =
+  ?username ?parse ?icon_url ?icon_emoji ?attachments text =
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: It would probably be more consistent to default `attachments` to `[]` and then `definitely_add` them, so there is no ambiguity and different behaviour between `Some []` and `None`.

Unless the Slack API does something weird when attachments are added but said list is empty. In such case it would probably still make sense to dispatch behaviour inside of this function and do not expose that in the signature.

File: src/slacko.ml (hunk 10 above)
Comment (jerith) [inline src/slacko.ml]: Rather than adding more parameters to `chat_post_message` (and potentially `chat_update`, which doesn't currently accept attachments, etc.) would it make sense to have a bigger `message` type that includes all the icons and attachments and whatnot instead of treating it as a string?

If so, I think that's a big enough change to be worth putting in its own PR after this one.

File: src/slacko.ml (hunk 3 above)
Comment (jerith) [inline src/slacko.ml]: Done.

**Comment by jerith** [comment]
Ready for the next round of review. :-)

File: src/slacko.ml (hunk 10 above)
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: There is a line between wrapping the Slack API and creating something (potentially more sensible) on top of it that we should keep in mind. Though in this case I would agree, it's a good idea for a later PR.

File: src/slacko.mli [hunk 13]
```diff
@@ -149,6 +149,14 @@ type message_length_error = [
   | `Msg_too_long
 ]
 
+(** When posting a message with attachments, you may receive this error if you
```
Comment (Leonidas-from-XIV) [inline src/slacko.mli]: I am a bit confused by this docstring, because it seems to state:

 * The maximum is 100 attachments
 * The maximum is 20 attachments

Comment (jerith) [inline src/slacko.mli]: I'll reword that. The docs state that you'll only get an error if you post more than 100 attachments, but that you shouldn't post more than 20.
